### ISSUE TITLE
Blocking commands do not respond to SIGTERM

### ISSUE THREAD

**Issue by taka-oyama** [phpredis/phpredis#2348]
### Expected behaviour

Blocking commands like `XREAD` (with `block = 0`) and `BLPOP` terminates when SIGTERM is received.

### Actual behaviour

Blocking commands do not respond to SIGTERM.

### I'm seeing this behaviour on
- OS: Alpine Linux 3.17.0
- Redis: 6
- PHP: 8.1
- phpredis: 5.3.7

### Steps to reproduce, backtrace or example script

1. Run the following script in CLI.

```
<?php

$redis = new Redis();
$redis->connect('redis', 6379);
$redis->xRead(['stream' => '0-0'], -1, 0);
```

2. Open another shell and call `kill -TERM {script pid}`

### I've checked
- [x] There is no similar issue from other users
- [x] Issue isn't fixed in `develop` branch

**Comment by michael-grunder** [comment]
I'm not able to replicate this.

I tried both on my Ubuntu box and using a Docker alpine container and I can terminate the script with `kill -TERM <pid>`

**Comment by taka-oyama** [comment]
Hi, thanks for the feed back. 

I forgot to mention that this was running under supervisord as a laravel command.

I tried running this again directly and it does indeed terminate properly, so there must be something happening between the script and supervisord.

Apologies for the confusion.

**Comment by taka-oyama** [comment]
I dug further and found that this is only reproducible when `pcntl_signal` is used to register a custom handler.

Could you please try again with the following code?

```
<?php

pcntl_async_signals(true);

pcntl_signal(SIGTERM, function () {
    echo 'terminate';
    exit(0);
});

$redis = new Redis();
$redis->connect('redis');
$redis->xRead(['stream' => '0-0'], -1, 0);
```

**Comment by michael-grunder** [comment]
I get the same behavior if I use PHP's sockets directly:

```php
<?php

printf("PID: %d\n", getmypid());

/* Note that true/false here doesn't make any difference */
pcntl_async_signals(false);

pcntl_signal(SIGTERM, function() {
    echo 'terminate';
    exit(0);
});

// address and port to connect
$host = '127.0.0.1';
$port = 6379;

// creating a socket
$socket = socket_create(AF_INET, SOCK_STREAM, SOL_TCP);
assert($socket !== false);

// connecting the socket
$result = socket_connect($socket, $host, $port);
assert($result !== false);

// message to be sent
$message = "XREAD BLOCK 0 STREAMS stream 0-0\r\n";
$result = socket_write($socket, $message, strlen($message));
assert($result == strlen($message));

// setting the socket option to never timeout
socket_set_option($socket, SOL_SOCKET, SO_RCVTIMEO, array('sec' => 0, 'usec' => 0));
while ($out = socket_read($socket, 2048)) {
    echo $out;
}

echo "Exiting...\n";
```

I didn't do a deep dive into the internals, but basically what's happening is that when PHP is on a blocking read like this, it uses special macros to indicate it is in a "critical section".  What this means specifically for signal handling, is that instead of being handled immediately, they are appended to a deferral list.

These deferred signals are then "replayed" once the critical section is done.  Unfortunately, in this case that will never happen, because PHP will happily restart the `read` syscall after appending to the deferral list.

I'm not aware of any idiomatic way to handle this internally.  One possible solution would be instead of sending SIGTERM to the pid, sending a special message to the stream itself which just means "break".  Alternitavely you could use a non-infinite timeout in a loop.  After a timeout, your signal would be replayed and you can then exit.


```php
<?php

pcntl_async_signals(true);
pcntl_signal(SIGTERM, function() {
    echo 'going bye-bye';
    exit(0);
});

$redis = new Redis();
$redis->connect('127.0.0.1', 6379);

while (true) {
    $messages = $redis->xRead(['stream' => '0-0'], -1, 1000.00);
    if ( ! $messages) {
        printf("[%d] No messages, trying again...\n", time());
        continue;
    }

    print_r($messages);
    $redis->del('stream');
}
```

**Comment by taka-oyama** [comment]
Hi, thanks for looking into the issue.
I found a similar issue on php.net and it doesn't seem like they're going to fix it anytime soon...

https://bugs.php.net/bug.php?id=49340

> Alternitavely you could use a non-infinite timeout in a loop. After a timeout, your signal would be replayed and you can then exit.

I will go with this solution for now.
